src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+        {
+            m_data[i].init_shape(proj);
+        }
+    }
+
+protected:
+    Tensor() : m_data(nullptr) {}
+    void init_shape(const Shape<N>& shape)
+    {
+        assert(!m_data);
+        base_type::set_size(shape.dim[0]);
+        m_data = new SubTensor[base_type::size()];
+        init_shape_alloced(shape);
+    }
+
+public:
+    Tensor(const Shape<N>& shape) :
+        TensorBase<N, T> (shape.dim[0]),
+        m_data(new SubTensor[base_type::size()])
+    {
+        init_shape_alloced(shape);
+    }
+
+    ~Tensor()
+    {
+        delete[] m_data;
+    }
+
+    Tensor& operator=(Tensor&& other) {
+        base_type::operator=(other);
+        m_data = other.m_data;
+        return *this;
+    }
+
+    //! access slice with index @p i
+    Tensor<N-1, T>& operator[](tensor_size_t i) { return m_data[i]; }
+    const Tensor<N-1, T>& operator[](tensor_size_t i) const { return m_data[i]; }
+
+    bool operator==(const Tensor<N, T>& other) const
+    {
+        bool equal = base_type::operator==(other);
+        for(tensor_size_t i = 0; equal && i < base_type::size(); ++i)
+        {
+            equal = m_data[i].operator==(other.m_data[i]);
+        }
+        return equal;
+    }
+
+    bool operator!=(const Tensor<N, T>& other) const {
+        return !operator==(other); }
+
+    void swapWith(Tensor<N,T>& other)
+    {
+        TensorBase<N, T>::swapWith(other);
+        std::swap(m_data, other.m_data);
+    }
+
+    class TensorIterator
+    {
+        using sub_type = Tensor<N-1, T>;
+        sub_type* sub;
+    public:
+        TensorIterator() = default;
+        TensorIterator(sub_type* sub) : sub(sub) {}
+        TensorIterator operator++() { ++sub; return *this; }
+        bool operator!=(TensorIterator& other) const { return sub != other.sub; }
+        sub_type& operator*() { return *sub; }
+    };
+    TensorIterator begin() const { return TensorIterator(m_data); }
+    TensorIterator end() const { return TensorIterator(m_data + base_type::size()); }
+
+    /*
+        testing only
+     */
+    tensor_size_t debug_set_data_using_deep_copy(const T* new_data)
+    {
+        tensor_size_t consumed = 0;
+        for(Tensor<N-1, T>& sub : *this)
+        {
+            consumed += sub.debug_set_data_using_deep_copy(new_data + consumed);
+        }
+        return consumed;
+    }
+
+    Shape<N> debug_get_shape() const {
+        if(base_type::size()) {
+            // we do not store the whole shape, but only our own size,
+            // so calculate it iterting
+            return m_data[0].debug_get_shape().prepend_dim(base_type::size());
+        }
+        else {
+            // we can not compute the shape from m_data[0], because
+            // m_data has size 0 - so set it all to 0
+            Shape<N> res;
+            for(tensor_size_t i = 0; i < N; ++i) res.dim[i] = 0;
+            return res;
+        }
+    }
+};
+
+//! Tensor class for dimension 1
```

Comment:
More of a cosmetic note, but in the future we might want to refactor some of these classes into their own headers within the Containers dir.